extstore.c

```diff
@@ -253,15 +253,32 @@ void *extstore_init(struct extstore_conf_file *fh, struct extstore_conf *cf,
     e->page_size = cf->page_size;
     uint64_t temp_page_count = 0;
     for (f = fh; f != NULL; f = f->next) {
-        f->fd = open(f->file, O_RDWR | O_CREAT | O_TRUNC, 0644);
+        f->fd = open(f->file, O_RDWR | O_CREAT, 0644);
         if (f->fd < 0) {
             *res = EXTSTORE_INIT_OPEN_FAIL;
 #ifdef EXTSTORE_DEBUG
-            perror("open");
+            perror("extstore open");
 #endif
             free(e);
             return NULL;
         }
+        // use an fcntl lock to help avoid double starting.
+        struct flock lock;
+        lock.l_type = F_WRLCK;
+        lock.l_start = 0;
+        lock.l_whence = SEEK_SET;
+        lock.l_len = 0;
+        if (fcntl(f->fd, F_SETLK, &lock) < 0) {
```

Comment:
lockf() seems to do the same thing in a more concise way, but if you're worried about portability, this would be better.

Also, reading the docs, fcntl() seems preferrable to flock() for our case here.

Reply:
portability is a concern yes. this was the oldest method I could find :)